Combine the WebExtension constructors to reduce duplication.

xeenon opened the pull request:
#### 1c8031fb856c3700466ff6a5543e201831f62e80
<pre>
Combine the WebExtension constructors to reduce duplication.
<a href="https://webkit.org/b/280909">https://webkit.org/b/280909</a>
<a href="https://rdar.apple.com/137293002">rdar://137293002</a>

Reviewed by Brian Weinstein.

Combine the bundle and resource URL WebExtension constructors, since there is a lot of overlap and clients
might need to provide different values. Also decouple resource validation from having a bundle set, we can
only do validation if the bundle and resource URL are equal.

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(+[WKWebExtension extensionWithAppExtensionBundle:completionHandler:]): Use the combined constructor.
(+[WKWebExtension extensionWithResourceBaseURL:completionHandler:]): Ditto.
(-[WKWebExtension _initWithAppExtensionBundle:error:]): Ditto.
(-[WKWebExtension _initWithResourceBaseURL:error:]): Ditto.
(-[WKWebExtension _initWithAppExtensionBundle:resourceBaseURL:error:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension): Combined the bundle and resource URL WebExtension constructors.
(WebKit::WebExtension::validateResourceData): Return early if m_shouldValidateResourceData is false.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: <a href="https://commits.webkit.org/284716@main">https://commits.webkit.org/284716@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a0ce22609569c0cdb4710fb6818ad91f8171b4e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70284 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49690 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74373 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/21462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21302 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/21462 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73350 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45221 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60598 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/36173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41880 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19823 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63809 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18377 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76092 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17614 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14552 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60664 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63362 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11398 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5025 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10755 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45492 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46566 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/47843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46308 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -122,44 +122,39 @@
-    }
-}
+#if PLATFORM(MAC)
+    m_shouldValidateResourceData = m_bundle ? [m_resourceBaseURL isEqual:bundleResourceURL] : false;
```
Nit: This could just be:

`m_shouldValidateResourceData = m_bundle && [m_resourceBaseURL isEqual:bundleResourceURL]`

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
```diff
@@ -93,24 +93,7 @@ - (instancetype)initWithAppExtensionBundle:(NSBundle *)appExtensionBundle error:
+    return [self _initWithAppExtensionBundle:appExtensionBundle resourceBaseURL:nil error:error];
 }
 
 // FIXME: Remove after Safari has adopted new methods.
```
not related to this change, but what's the bug/radar tracking this?

xeenon committed:
Combine the WebExtension constructors to reduce duplication.
https://webkit.org/b/280909
rdar://137293002

Reviewed by Brian Weinstein.

Combine the bundle and resource URL WebExtension constructors, since there is a lot of overlap and clients
might need to provide different values. Also decouple resource validation from having a bundle set, we can
only do validation if the bundle and resource URL are equal.

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(+[WKWebExtension extensionWithAppExtensionBundle:completionHandler:]): Use the combined constructor.
(+[WKWebExtension extensionWithResourceBaseURL:completionHandler:]): Ditto.
(-[WKWebExtension _initWithAppExtensionBundle:error:]): Ditto.
(-[WKWebExtension _initWithResourceBaseURL:error:]): Ditto.
(-[WKWebExtension _initWithAppExtensionBundle:resourceBaseURL:error:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::WebExtension): Combined the bundle and resource URL WebExtension constructors.
(WebKit::WebExtension::validateResourceData): Return early if m_shouldValidateResourceData is false.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: https://commits.webkit.org/284716@main
